tests/integration/helpers.py

```diff
@@ -0,0 +1,199 @@
+        else:
+            if result := func():
+                return cast(R, result)
+        time.sleep(check_interval)
+
+    # final check before raising TimeoutError.
+    if is_awaitable:
+        if result := await cast(Awaitable, func()):
+            return result
+    else:
+        if result := func():
+            return cast(R, result)
+    raise TimeoutError()
+
+
+IMAGE_TO_TAG = {"jammy": "22.04", "noble": "24.04"}
+
+
+def _create_metadata_tar_gz(image: str, tmp_path: Path) -> Path:
+    """Create metadata.tar.gz contents.
+
+    Args:
+        image: The ubuntu LTS image name.
+        tmp_path: Temporary dir.
+    """
+    # Create metadata.yaml
+    template = Template(
+        Path("tests/integration/testdata/metadata.yaml.tmpl").read_text(encoding="utf-8")
+    )
+    metadata_contents = template.substitute(
+        {"arch": platform.machine(), "tag": IMAGE_TO_TAG[image], "image": image}
+    )
+    meta_path = tmp_path / "metadata.yaml"
+    meta_path.write_text(metadata_contents, encoding="utf-8")
+
+    # Pack templates/ and metada.yaml
+    templates_path = Path("tests/integration/testdata/templates")
+    metadata_tar = tmp_path / Path("metadata.tar.gz")
+
+    with tarfile.open(metadata_tar, "w:gz") as tar:
+        tar.add(meta_path, arcname=meta_path.name)
+        tar.add(templates_path, arcname=templates_path.name)
+
+    return metadata_tar
+
+
+# This is a workaround until https://github.com/canonical/pylxd/pull/577 gets merged.
+def _post_vm_img(
+    client: Client,
+    image_data: bytes,
+    metadata: bytes | None = None,
+    public: bool = False,
+) -> Image:
+    """Create an LXD VM image.
+
+    Args:
+        client: The LXD client.
+        image_data: Image qcow2 (.img) file contents in bytes.
+        metadata: The metadata.tar.gz contents in bytes.
+        public: Whether the image should be publicly available.
+    """
+    headers = {}
+    if public:
+        headers["X-LXD-Public"] = "1"
+
+    if metadata is not None:
+        # Image uploaded as chunked/stream (metadata, rootfs)
+        # multipart message.
+        # Order of parts is important metadata should be passed first
+        files = collections.OrderedDict(
+            {
+                "metadata": ("metadata", metadata, "application/octet-stream"),
+                # rootfs is container, rootfs.img is VM
+                "rootfs.img": ("rootfs.img", image_data, "application/octet-stream"),
+            }
+        )
+        data = MultipartEncoder(files)
+        headers.update({"Content-Type": data.content_type})
+    else:
+        data = image_data
+
+    response = client.api.images.post(data=data, headers=headers)
+    operation = client.operations.wait_for_operation(response.json()["operation"])
+    return Image(client, fingerprint=operation.metadata["fingerprint"])
+
+
+def create_lxd_vm_image(lxd_client: Client, img_path: Path, image: str, tmp_path: Path) -> Image:
+    """Create LXD VM image.
```

Comment:
Perhaps a bit more explanation in the docstring of what is done with the supplied image to convert it to LXD.

Reply:
Added, thank you!